### PR TITLE
Endpoint for user configuration (#6)

### LINKED ISSUE #6: Add endpoint to store user configuration
* Add an HTML form to get user input from the web app.
* Add an API handler in the backend to get the user input.
* Make sure API key is ciphered (never send credentials in plain text) using a token.
* Store the configuration in datastore (https://cloud.google.com/datastore/docs)

### PR DESCRIPTION
**Pull request by karpp**
Closes #6

### CODE REVIEW

File: default/main.py [hunk 1]
```diff
@@ -1,14 +1,28 @@
...
 
 templates = Jinja2Templates(directory='templates')
 
+datastore_client = datastore.Client()
+
 
 @app.get('/')
 async def index(request: Request):
-    return templates.TemplateResponse("index.html", {'request': request, 'content': 'Hello, world! =)'})
+    return templates.TemplateResponse("form.html", {'request': request, 'content': 'Hello, world! =)'})
```
Comment (mgmacias95) [inline default/main.py:17]: Change the hello world message

File: default/requirements-dev.txt [hunk 2]
```diff
@@ -1,27 +1,43 @@
 aiofiles==0.5.0
 astroid==2.4.2
 attrs==19.3.0
+cachetools==4.1.1
 certifi==2020.6.20
 chardet==3.0.4
 click==7.1.2
 fastapi==0.59.0
+Flask==1.1.2
+Flask-WTF==0.14.3
```
Comment (mgmacias95) [inline default/requirements-dev.txt:10]: are you finally using this?

File: default/tests/testfiles/index.html [hunk 3]
```diff
@@ -2,9 +2,19 @@
...
+            <form method="POST" action="/userdata/">
+                <label for="apikey">API Key</label>
+                <input type="password" name="apikey" id="apikey">
+                <label for="webhook">Webhook URL</label>
+                <input type="url" name="webhook" id="webhook">
+                <input type="submit" value="Go">
+            </form>
+        </div>
     </body>
 </html>
```
Comment (niedziol) [inline default/tests/testfiles/index.html:20]: In few files there are new new lines at the end.
When we'll have CI it'll probably show warnings about it, but for now it's our problem.

File: default/tests/test_main.py [hunk 4]
```diff
@@ -10,3 +10,22 @@ def test_index():
...
+    assert response.status_code == 200
+    with open('tests/testfiles/userdata.html', 'r') as correct:
+        assert response.text == correct.read()
+    
+    query = datastore_client.query(kind='userdata')
+    query.add_filter('apikey', '=', 'test_apikey')
+    result = list(query.fetch())
+    assert result != []
+    key = result[0].key
+    datastore_client.delete(key)
```
Comment (niedziol) [inline default/tests/test_main.py:31]: You also need to test cases where there are no params, or params are wrong.

File: default/tests/test_main.py [hunk 5]
```diff
@@ -1,7 +1,7 @@
 from fastapi import FastAPI
 from fastapi.testclient import TestClient
 
-from default.main import app
+from default.main import app, datastore_client
```
Comment (mgmacias95) [inline default/tests/test_main.py:4]: how does the normal datastore_client work when testing? does it connect to prod?

File: default/requirements.txt [hunk 6]
```diff
@@ -1,4 +1,6 @@
 aiofiles==0.5.0
 fastapi==0.59.0
+google-cloud-datastore==1.13.1
```
Comment (mgmacias95) [inline default/requirements.txt:3]: Take a look at python-ndb (https://github.com/googleapis/python-ndb). I prefer the interface ndb provides, defining the model as a class.

File: default/tests/test_main.py [hunk 7]
```diff
@@ -10,3 +10,22 @@ def test_index():
     assert response.status_code == 200
     with open('tests/testfiles/index.html', 'r') as correct:
         assert response.text == correct.read()
+
+
+def test_userdata():
```
Comment (mgmacias95) [inline default/tests/test_main.py:15]: Add more testing:
* What happens if there are missing parameters
* What happens if any of the parameters is incorrect
* etc

File: default/tests/testfiles/index.html (hunk 3 above)
Comment (karpp) [reply default/tests/testfiles/index.html:20]: But it's for unittests and when I request page from webapp, it comes without newlines, and I check it for exact match

Comment (niedziol) [reply default/tests/testfiles/index.html:20]: Ok, I thought that there'll be newline at the ending, but probably I was wrong.

File: default/main.py [hunk 8]
```diff
@@ -1,14 +1,27 @@
...
-    return templates.TemplateResponse("index.html", {'request': request, 'content': 'Hello, world! =)'})
+    return templates.TemplateResponse("form.html", {'request': request, 'content': 'Save your credentials'})
+
+
+@app.post('/userdata/')
+async def userdata(request: Request, apikey: str = Form(...), webhook: str = Form(...)):
+    with client.context():
+        task = models.Userdata(apikey=apikey, webhook=webhook)
+        task.put()
+    return templates.TemplateResponse("index.html", {'request': request, 'content': 'Your answer successfully saved'})
```
Comment (niedziol) [inline default/main.py:27]: Your answer ***was*** successfully saved

File: default/requirements-dev.txt [hunk 9]
```diff
@@ -1,27 +1,43 @@
...
 attrs==19.3.0
+cachetools==4.1.1
 certifi==2020.6.20
 chardet==3.0.4
 click==7.1.2
 fastapi==0.59.0
+google-api-core==1.21.0
+google-auth==1.19.0
+google-cloud-core==1.3.0
+google-cloud-datastore==1.13.1
```
Comment (mgmacias95) [inline default/requirements-dev.txt:12]: are you still using this one?

Comment (karpp) [reply default/requirements-dev.txt:12]: Yes, google-cloud-ndb depends on it

**Pull request merged**